### PR TITLE
Add BA code to warnings

### PR DESCRIPTION
**Pull request by rouille**
### Purpose
Add BA code to printouts.

### What the code is doing
Whenever a plant-level data frame is printed out when running the pipeline, the ba code associated to each plant is retrieved and inserted in the printed data frame. Existing functions are used to create a helper function that return a dictionary mapping the EIA plant id to the BA code.

### Testing
Running the pipeline

### Where to look
* Move helper functions into a new `oge.helpers` module
* Fix `pandas` related warning throughout the code
* Fix a `if` statement in the `check_for_complete_monthly_timeseries` in the `oge.validation` module

### Usage Example/Visuals
```bash
    plant_id_eia  gross_generation_mwh  net_generation_mwh  annual_plant_ratio ba_code
39         54096               18039.0            162310.0            8.997727    SOCO
27         10567               17734.0            134118.2            7.562772    ISNE
10          2831                1913.0              7158.4            3.741976     PJM
67         55470             1200399.0           3786786.0            3.154606    ERCO
79         65372                7427.0             17138.7            2.307621    ERCO
68         55672              164399.0            285386.4            1.735938    SOCO
58         55211             1558631.0           2561209.0            1.643243    ISNE
9           2712             3346069.0           5460053.1            1.631781    CPLE
25          8059              177444.0            284855.0            1.605323    SWPP
37         52168               57475.0             89516.3            1.557482    NYIS
73         56565             2008929.0           3121630.0            1.553878    SWPP
3            689              878403.0           1354941.0            1.542505     TAL
48         55068              434128.0            669598.0            1.542398    ISNE
65         55406             1651462.0           2544639.0            1.540840    SOCO
14          7302             7956453.0          12240582.0            1.538447     FPC
2            676             1522950.0           2342083.8            1.537860    FMPP
78         62548               22691.0             34892.2            1.537711    ERCO
64         55372             1740686.0           2674207.0            1.536295    HGMA
52         55089             3710150.0           5698863.0            1.536020    MISO
66         55457             1497373.0           2298880.0            1.535275    SWPP
54         55149             3728636.0           5722994.0            1.534876    ISNE
13          4937             2051426.0           3143980.0            1.532583    ERCO
71         55976             3869629.0           5914124.0            1.528344     PJM
80         65978             2546859.0           3880981.9            1.523831     FPL
43         54537              644497.0            981634.0            1.523101    PSEI
55         55154             2066741.0           3141227.0            1.519894    ERCO
74         56806             1615115.0           2453095.0            1.518836    ERCO
17          7699              838072.0           1266028.0            1.510643     FPC
22          7873             5102750.0           7698823.8            1.508760     TEC
5           1393             2180774.0           3274613.0            1.501583    MISO
51         55088             2964286.0           4447878.7            1.500489    MISO
24          7985              984832.0           1475295.0            1.498017    MISO
57         55207                9007.0             13467.4            1.495215    PSCO
49         55079              771685.0           1152563.0            1.493567    ISNE
61         55239             2107041.0           3137651.0            1.489127     PJM
44         54592                 368.0               547.0            1.486413    NYIS
63         55337             3456386.0           5136582.0            1.486114     PJM
30         10776                2722.0              4027.0            1.479427    CISO
16          7380             2542785.0           3757745.0            1.477807     SEC
56         55170             2324179.0           3433772.0            1.477413    ISNE
8           2539             2887468.0           4262510.0            1.476210    NYIS
70         55710             2988397.0           4405161.0            1.474088     PJM
23          7946             1591560.0           2333887.0            1.466415    SOCO
20          7846             3324659.0           4871811.0            1.465357     JEA
77         57978              575491.0            840306.0            1.460155    CISO
59         55212             1781983.0           2598161.0            1.458017    ISNE
45         54640              260353.0            379559.0            1.457863     PJM
15          7314              459746.0            668301.0            1.453631    NYIS
0            533             2681877.0           3895666.0            1.452589    SOCO
47         55043              289727.0            418180.6            1.443361     DUK
36         51030              530813.0            765028.0            1.441238    ISNE
60         55231             2477537.0           3556261.0            1.435402     PJM
21          7870              297757.0            426767.0            1.433273    PSEI
62         55295             1103518.0           1579922.0            1.431714    CISO
38         54041              135935.0            194505.3            1.430870    NYIS
11          3236              693275.0            991141.0            1.429651    ISNE
42         54476              321713.0            458958.0            1.426607    PSEI
41         54324              496544.0            708046.0            1.425948    ISNE
12          3295             1412631.0           2007514.1            1.421117    SCEG
6           2398             1464621.0           2065411.0            1.410202     PJM
4           1007             1105621.0           1551750.0            1.403510    MISO
75         57564              127570.0            176967.0            1.387215    CISO
18          7826             3235073.0           4474654.1            1.383169     DUK
69         55706             2495062.0           3425619.0            1.372959     TVA
1            607               62435.0             85612.2            1.371221    PSEI
53         55096               93827.0            128152.4            1.365837    MISO
29         10726              171698.0            233904.0            1.362299    ISNE
50         55086              516135.0            699287.0            1.354853    ERCO
40         54131               41076.0             55348.0            1.347454    NYIS
31         50002               42188.0             56796.0            1.346260    ISNE
26         10307              159406.0            211088.0            1.324216    ISNE
72         56151              164123.0            216594.0            1.319705    SWPP
7           2493             2300590.0           3011081.0            1.308830    NYIS
33         50744                2926.0              3778.0            1.291183    NYIS
19          7834             4413919.0           5688275.2            1.288713      SC
32         50555                1281.0              1643.0            1.282592     PJM
34         50949               10974.0             14069.6            1.282085     TEC
46         54785              909303.0           1156415.1            1.271760     PJM
35         50978              120494.0            152666.0            1.267001    NYIS
28         10633              536945.0            677731.0            1.262198     PJM
76         57703              225755.0            282814.0            1.252747    WACM

```

### Review estimate
20min

### Future work
Flag anomalous values in CEMS input timeseries.

### Checklist
- [x] Update the documentation to reflect changes made in this PR
- [x] Format all updated python files using `black`
- [x] Clear outputs from all notebooks modified
- [x] Add docstrings and type hints to any new functions created

### CODE REVIEW

File: src/oge/data_pipeline.py [hunk 1]
```diff
@@ -78,6 +79,8 @@ def main(args):
     args = get_args()
     year = args.year
 
+    os.environ["OGE_DATA_YEAR"] = "2022"
```
Comment (grgmiller) [inline src/oge/data_pipeline.py:82]: I don't think we want to manually specify the year here - if anything, this should always be derived from `args.year`

File: src/oge/emissions.py [hunk 2]
```diff
@@ -674,6 +686,7 @@ def calculate_generator_nox_ef_per_unit_from_boiler_type(gen_fuel_allocated, yea
         gen_nox_factors.loc[
             gen_nox_factors["emission_factor"].isna(),
             [
+                "plant_id_eia",
```
Comment (grgmiller) [inline src/oge/emissions.py:689]: For this check, we do not want to add plant IDs or BA codes - we only care about generic prime movers across all plants.

File: src/oge/emissions.py [hunk 3]
```diff
@@ -691,6 +704,14 @@ def calculate_generator_nox_ef_per_unit_from_boiler_type(gen_fuel_allocated, yea
         )
     )
     if len(missing_nox_efs) > 0:
+        missing_nox_efs.insert(
```
Comment (grgmiller) [inline src/oge/emissions.py:707]: We do not need BA codes on this check since it is not a plant-specific validation check

File: src/oge/emissions.py [hunk 4]
```diff
@@ -705,6 +726,7 @@ def calculate_generator_nox_ef_per_unit_from_boiler_type(gen_fuel_allocated, yea
         gen_nox_factors.loc[
             gen_nox_factors["emission_factor"].isna(),
             [
+                "plant_id_eia",
```
Comment (grgmiller) [inline src/oge/emissions.py:729]: See above: For this check, we do not want to add plant IDs or BA codes - we only care about generic prime movers across all plants.

File: src/oge/emissions.py [hunk 5]
```diff
@@ -721,13 +743,22 @@ def calculate_generator_nox_ef_per_unit_from_boiler_type(gen_fuel_allocated, yea
...
     )
+
     if len(missing_nox_efs) > 0:
         logger.warning(
             """
             After filling with PM-fuel factors, NOx emission factors are still missing for the following boiler types.
             An emission factor of zero will be used for these boilers.
             Missing factors for FC prime movers are currently expected."""
         )
+        missing_nox_efs.insert(
```
Comment (grgmiller) [inline src/oge/emissions.py:754]: See above: We do not need BA codes on this check since it is not a plant-specific validation check

File: src/oge/emissions.py [hunk 6]
```diff
@@ -1405,6 +1446,7 @@ def calculate_generator_so2_ef_per_unit_from_boiler_type(gen_fuel_allocated, yea
         gen_so2_factors.loc[
             gen_so2_factors["emission_factor"].isna(),
             [
+                "plant_id_eia",
```
Comment (grgmiller) [inline src/oge/emissions.py:1449]: See above comment, we should not add plant ID here

File: src/oge/emissions.py [hunk 7]
```diff
@@ -1419,11 +1461,20 @@ def calculate_generator_so2_ef_per_unit_from_boiler_type(gen_fuel_allocated, yea
             ]
         )
     )
+
     if len(missing_so2_efs) > 0:
         logger.warning(
             "SO2 emission factors are missing for the following boiler types. A prime mover-fuel level factor will be used if available."
         )
         logger.warning("Missing factors for FC prime movers are currently expected")
+        missing_so2_efs.insert(
```
Comment (grgmiller) [inline src/oge/emissions.py:1470]: See above comment, we do not need BA codes here

File: src/oge/emissions.py [hunk 8]
```diff
@@ -1434,6 +1485,7 @@ def calculate_generator_so2_ef_per_unit_from_boiler_type(gen_fuel_allocated, yea
         gen_so2_factors.loc[
             gen_so2_factors["emission_factor"].isna(),
             [
+                "plant_id_eia",
```
Comment (grgmiller) [inline src/oge/emissions.py:1488]: See above comment, we do not want plant IDs here

File: src/oge/emissions.py [hunk 9]
```diff
@@ -1448,11 +1500,20 @@ def calculate_generator_so2_ef_per_unit_from_boiler_type(gen_fuel_allocated, yea
             ]
         )
     )
+
     if len(missing_so2_efs) > 0:
         logger.warning(
             "SO2 emission factors are missing for the following boiler types. An emission factor of zero will be used for these boilers."
         )
         logger.warning("Missing factors for FC prime movers are currently expected")
+        missing_so2_efs.insert(
```
Comment (grgmiller) [inline src/oge/emissions.py:1509]: See above comment, we do not need BA codes here.

File: src/oge/helpers.py
Comment (grgmiller) [inline src/oge/helpers.py:1]: For this file, how did you choose which functions got moved here vs what stayed in `data_cleaning`?

Comment (grgmiller) [inline src/oge/helpers.py:1]: Also, is `map_eia_plant_id_to_ba_code` the only new/modified function in this file, or were any other functions added or modified?

File: src/oge/helpers.py [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+        validate="m:1",
+    )
+    plant_ba.update(
+        pd.DataFrame({"ba_code_physical": plant_ba["ba_code_physical_map"].to_list()})
+    )
+
+    return plant_ba
+
+
+def map_eia_plant_id_to_ba_code():
```
Comment (grgmiller) [inline src/oge/helpers.py:272]: It seems like this will work, but this solution (including the new `utility` module) seems to add a lot of unneeded complexity just to  avoid running `create_plant_ba_table()` (which only takes 1 to 1.5 seconds to load a pretty small dataframe) and merging it into the dataframe we're validating a handful of times. 

It seems like it would have been far simpler to just run
```
df = df.merge(create_plant_ba_table(year)[["plant_id_eia,"ba_code"]], how="left", on="plant_id_eia", validate="m:1")
```
and pass `year` to wherever we needed it (or access it though a global variable.

File: src/oge/helpers.py [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+    """Maps EIA plant id to BA code"""
+    year = (
+        2022 if os.getenv("OGE_DATA_YEAR") is None else int(os.getenv("OGE_DATA_YEAR"))
```
Comment (grgmiller) [inline src/oge/helpers.py:275]: the year should always be accessed from `args.year`

File: src/oge/load_data.py [hunk 12]
```diff
@@ -221,7 +221,9 @@ def update_epa_to_eia_map(cems_df: pd.DataFrame):
     )
 
     # update the eia plant ids
-    cems_df["plant_id_eia"].update(cems_df["plant_id_eia_manual"])
+    cems_df.update(
+        pd.DataFrame({"plant_id_eia": cems_df["plant_id_eia_manual"].to_list()})
```
Comment (grgmiller) [inline src/oge/load_data.py:225]: Does this approach preserve the NA values in the column being sent to list so that we are only overwriting non-NA values? I just want to make sure we are not accidentally replacing data we do not intend. (This applies to all instances of this solution throughout the PR)

File: src/oge/validation.py [hunk 13]
```diff
@@ -320,6 +362,9 @@ def check_non_missing_cems_co2_values_unchanged(cems_original, cems):
 
 def check_removed_data_is_empty(cems):
     """Checks that the rows removed by `data_cleaning.remove_cems_with_zero_monthly_data()` don't actually contain non-zero data"""
+    print("############# Begin validation.py:353 #############")
```
Comment (grgmiller) [inline src/oge/validation.py:365]: These print statements should be removed

**Review by grgmiller — COMMENTED** [review]
At a high level, I'm wondering if the proposed solution here is a bit over-engineered and adding unnecessary complexity. Maybe I don't fully understand what problem the memory caching is solving, but it seems like the addition of a ba code to these warning messages could have been addressed with a simpler solution like we discussed like: 
`df = df.merge(create_plant_ba_table(year)[["plant_id_eia,"ba_code"]], how="left", on="plant_id_eia", validate="m:1")`.
If we want to avoid re-loading this dataframe multiple times, one other potential solution could have been to load it once at the top of `validation.py` and have it accessed as a global variable in each validation function that uses it. 

I appreciate all the work to address the warning messages we've been seeing in pandas, as well as splitting out `data_cleaning` into smaller chunks by adding `helpers` (although see my comment about ensuring we are consistent about what functions end up where).

A couple other high-level comments:
- If we end up restructuring the code and/or adding new modules, we'll need to make sure that the readme is updated
- Whenever we are adding ba_codes to a dataframe, we should make sure that we are only ever adding this to "throwaway" dfs that are not passed further in the data pipeline; otherwise, we should be sure to drop the ba_code column from the df after the warning message.

File: src/oge/data_pipeline.py (hunk 1 above)
Comment (rouille) [reply src/oge/data_pipeline.py:82]: The `create_plant_ba_table`  function (now in the `oge.helpers` module) takes `year` as argument. Should it be set to 2022. I was trying to use the year that is used in the pipeline and that I cannot access in the `oge.helpers` module

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py:1]: I only moved what was needed for this PR. I believe other functions could be moved. As we discussed, the reason for creating this new module was the circular import I was encountering when trying to import `oge.data_cleaning` in other modules.

File: src/oge/helpers.py [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+    plant_attributes = plant_attributes.drop(columns=["plant_primary_fuel_cems"])
+    plant_attributes["data_availability"] = plant_attributes[
+        "data_availability"
+    ].cat.rename_categories(
+        {
+            "left_only": "eia_only",
+            "right_only": "cems_only",
+            "both": "cems_and_eia",
+        }
+    )
```
Comment (rouille) [inline src/oge/helpers.py:88]: Here is the fixed `pandas` warning

File: src/oge/helpers.py (hunk 10 above)
Comment (rouille) [reply src/oge/helpers.py:272]: Sure. I will do what you propose

File: src/oge/helpers.py (hunk 11 above)
Comment (rouille) [reply src/oge/helpers.py:275]: This will be gone as I will implement your solution

File: src/oge/load_data.py (hunk 12 above)
Comment (rouille) [reply src/oge/load_data.py:225]: I believe it does as the below example shows:
```python
>>> df = pd.DataFrame({'A': [1.0, 2.0, 3.0],
...                    'B': [400.0, 500.0, np.NaN]})
>>> df
     A      B
0  1.0  400.0
1  2.0  500.0
2  3.0    NaN
>>> df.update({"A": df["B"].tolist()})
>>> df
       A      B
0  400.0  400.0
1  500.0  500.0
2    3.0    NaN
>>> 

```

File: src/oge/validation.py (hunk 13 above)
Comment (rouille) [reply src/oge/validation.py:365]: I was still evaluating the outcome

**Comment by rouille** [comment]
> At a high level, I'm wondering if the proposed solution here is a bit over-engineered and adding unnecessary complexity. Maybe I don't fully understand what problem the memory caching is solving, but it seems like the addition of a ba code to these warning messages could have been addressed with a simpler solution like we discussed like: `df = df.merge(create_plant_ba_table(year)[["plant_id_eia,"ba_code"]], how="left", on="plant_id_eia", validate="m:1")`. If we want to avoid re-loading this dataframe multiple times, one other potential solution could have been to load it once at the top of `validation.py` and have it accessed as a global variable in each validation function that uses it.
> 
> This is all to say - if there is a compelling reason for caching in memory, let's do that, but that reason is not currently clear to me, and I'm hesitant to add more complexity to OGE if not needed.
> 
> I appreciate all the work to address the warning messages we've been seeing in pandas, as well as splitting out `data_cleaning` into smaller chunks by adding `helpers` (although see my comment about ensuring we are consistent about what functions end up where).
> 
> A couple other high-level comments:
> 
>     * If we end up restructuring the code and/or adding new modules, we'll need to make sure that the readme is updated
> 
>     * Whenever we are adding ba_codes to a dataframe, we should make sure that we are only ever adding this to "throwaway" dfs that are not passed further in the data pipeline; otherwise, we should be sure to drop the ba_code column from the df after the warning message.

I will implement your solution, remove the `oge.utility` module and update the README to list the new helper module.

File: src/oge/emissions.py [hunk 15]
```diff
@@ -542,16 +557,17 @@ def calculate_nox_from_fuel_consumption(
...
         "controlled_nox_mass_lb"
     ].fillna(
         gen_fuel_allocated["fuel_consumed_mmbtu"]
         * gen_fuel_allocated["controlled_annual_nox_ef_lb_per_mmbtu"]
     )
     # update the emision total using the controlled mass if available
-    gen_fuel_allocated["nox_mass_lb"].update(
-        gen_fuel_allocated["controlled_nox_mass_lb"]
+    gen_fuel_allocated.update(
+        pd.DataFrame({"nox_mass_lb": gen_fuel_allocated["controlled_nox_mass_lb"]})
```
Comment (grgmiller) [inline src/oge/emissions.py:570]: So this looks like a slightly different pattern than what Jeff was doing, which I think would be:
` gen_fuel_allocated.update({"nox_mass_lb": gen_fuel_allocated["controlled_nox_mass_lb"]}`

Does this need to be converted to a dataframe for this to work or would a dict work? Just out of curiosity.

File: src/oge/gross_to_net_generation.py [hunk 16]
```diff
@@ -64,16 +67,16 @@ def convert_gross_to_net_generation(cems, eia923_allocated, plant_attributes, ye
         cems["gross_generation_mwh"] * cems["annual_plant_ratio"]
     )
 
-    cems.loc[
-        cems["net_generation_mwh"].isna(), "gtn_method"
-    ] = "3_annual_subplant_shift_factor"
+    cems.loc[cems["net_generation_mwh"].isna(), "gtn_method"] = (
+        "3_annual_subplant_shift_factor"
```
Comment (grgmiller) [inline src/oge/gross_to_net_generation.py:71]: why is this getting reformatted to ("string")? Is this the standard ruff formatting?

In general it seems like there were a lot of formatting updates, which is surprising to me since the whole repo had already been formatted with ruff...

File: src/oge/helpers.py
Comment (grgmiller) [reply src/oge/helpers.py:1]: It would be nice to consistently move these around, although not sure if that is worth it right now or not.

File: src/oge/impute_hourly_profiles.py [hunk 17]
```diff
@@ -236,7 +236,7 @@ def select_best_available_profile(hourly_profiles):
         ] = profile
         # fill missing values with non-missing values from the filtered profile data
         hourly_profiles["profile"] = hourly_profiles["profile"].fillna(
-            hourly_profiles[profile]
+            hourly_profiles[profile].astype("float64")
```
Comment (grgmiller) [inline src/oge/impute_hourly_profiles.py:239]: Why is this change necessary?

File: src/oge/validation.py [hunk 18]
```diff
@@ -250,14 +315,33 @@ def check_for_orphaned_cc_part_in_subplant(subplant_crosswalk):
...
+        )
+        logger.warning(
+            "\n"
+            + orphaned_cc_parts.assign(
+                ba_code=[
+                    id2ba[i]
+                    for i in orphaned_cc_parts.index.get_level_values("plant_id_eia")
+                ]
+            )
+            # merge(
```
Comment (grgmiller) [inline src/oge/validation.py:329]: was this meant to be converted to merge? Otherwise why are we using a different format here?

**Review by grgmiller — APPROVED** [review]
Looks good overall - a few outstanding comments to address

File: src/oge/emissions.py (hunk 15 above)
Comment (rouille) [reply src/oge/emissions.py:570]: I will try. Keep you updated.

File: src/oge/gross_to_net_generation.py (hunk 16 above)
Comment (rouille) [reply src/oge/gross_to_net_generation.py:71]: It comes from the new 0.3.0 ruff

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py:1]: I agree. We can make a list of functions to move and tackle that in a new PR.

File: src/oge/validation.py (hunk 18 above)
Comment (rouille) [reply src/oge/validation.py:329]: It is a multi-index data frame and the merge operation strip the second index (subplant_id). The current operation conserve the multi indexing.

File: src/oge/impute_hourly_profiles.py (hunk 17 above)
Comment (rouille) [reply src/oge/impute_hourly_profiles.py:239]: I try to fix the following warning:
```
FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'residual_profile_filtered' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
```
but it doesn't work so I will remove it.

**Comment by rouille** [comment]
Here is the logfile. It has been compared against the one from the release and it looks good.
[data_pipeline.log](https://github.com/singularity-energy/open-grid-emissions/files/14464553/data_pipeline.log)

File: src/oge/emissions.py (hunk 15 above)
Comment (rouille) [reply src/oge/emissions.py:570]: It works. Chnging it.

**Review by grgmiller — APPROVED** [review]
Looks good to me

File: src/oge/emissions.py (hunk 15 above)
Comment (rouille) [reply src/oge/emissions.py]: Done

File: src/oge/impute_hourly_profiles.py (hunk 17 above)
Comment (rouille) [reply src/oge/impute_hourly_profiles.py]: Removed
